src/Observable.ts

```diff
@@ -35,19 +35,78 @@ export class Observable<T> extends BaseObservable<T> {
     }
 
     map<U>(action: (value: T) => U): Observable<U> {
-        throw new Error('not implemented')
+        return new Observable<U>(observer => {
+            this.subscribe(
+                value => {
+                    if (observer.closed) {
+                        return
+                    }
+
+                    let newValue
+                    try {
+                        newValue = action(value)
+                    } catch (e) {
+                        return observer.error(e)
+                    }
+                    return observer.next(newValue)
+                },
+                observer.error,
+                observer.complete
+            )
+        })
     }
 
     flatMap<U>(action: (value: T) => Observable<U>): Observable<U> {
         throw new Error('not implemented')
     }
 
     reduce<U>(reductor: (value: T, summary: U) => U): Observable<U> {
-        throw new Error('not implemented')
+        return new Observable<U>(observer => {
+            let _summary : U = null
+            this.subscribe(
+                value => {
+                    if (observer.closed) {
+                        return
+                    }
+
+                    let newValue
+                    try {
+                        newValue = _summary
+                            ? reductor(value, _summary)
+                            : value
+                        _summary = newValue
+                    } catch (e) {
+                        return observer.error(e)
+                    }
+                    return observer.next(newValue)
+                },
+                observer.error,
+                observer.complete
+            )
+        })
     }
 
     filter(filter: (value: T) => boolean): Observable<T> {
-        throw new Error('not implemented')
+        return new Observable<T>(observer => {
+            this.subscribe(
```

Comment:
jest połączenie zakończenia strumienia pochodnego w razie zakonczenia strumienia źródłowego

ale nie ma połączenia anulowania subskrypcji w razie żądania zakończenia strumienia pochodnego

::subscribe zwraca obiekt subskrypcji, który ma metodę unsubscribe.
funkcja podana w constructor Observable'a może zwrócić funkcję, która będzie uruchomiona, jeżeli skądś przyjdzie żądanie zakończenia tego nowego strumienia.

funkcja podana w constructor Observable'a powinna zwracać funkcję, która wywoła to unsubscribe